### ISSUE TITLE
get_fs_records behaviour when loop reaches btn_nkeys

### ISSUE THREAD

**Issue by memecode** [jivanpal/apfs-tools#4]
So I've been looking why a certain recent folder can't be listed by apfs-list and worked back to what I believe the issue is.

In the get_fs_records function there is a loop titled "Determine which entry in this node we should descend":
`for (desc_path[i] = 0;    desc_path[i] < node->btn_nkeys;    desc_path[i]++, toc_entry++)`

For my oid it reaches the end of the loop and desc_path[i] == node->btn_nkeys... so it exits the loop. However the toc_entry->v.off == 0... which doesn't reference a valid index for child_node_virt_oid... right? maybe?

I printf'd out the oid vs record_oid as it goes through the loop:
path 'Dunkirk (2017) 5.1 Dual Áudio 4K UtraHD By-LuanHarper' = 3307851
	get_fs_records:301 - [0] record_oid=1 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=181686 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=364079 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=388618 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=556537 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=691924 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=787838 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=893180 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=899653 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1037106 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1058869 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1189538 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1388293 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1534138 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1682289 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1832542 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1954405 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=1966472 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=2114638 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=2213200 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=2337265 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=2528369 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=2674253 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=2846735 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=3062975 oid=3307851 node->btn_nkeys=26
	get_fs_records:301 - [0] record_oid=3222652 oid=3307851 node->btn_nkeys=26
	get_fs_records:358 - *child_node_virt_oid=17592186044482 toc_entry->v.off=0
	get_btree_phys_omap_val:50 - oid=17592186044482

It almost gets up the value it's searching for but then runs out of entries. So in my limited knowledge of b-trees, maybe the value it's looking for is in the children of the last element? And it should descend down there to find it? Maybe?

Seems like an unhandled edge case. Or I could be barking up the wrong [b]tree here. Haha. Maybe you could have a brief look at the code and see if there is something missing?

**Comment by memecode** [comment]
I've had some luck with putting this in there:
https://github.com/memecode/apfs-tools/commit/36d0cee109df6c7c344d63955f8eae7c21e3ae21#diff-ed645dbf88753c45bcd7100d5e899dd3R130-R131

Basically if you walk off the end of the array, just use the last element. This is walking down the very right(?) edge of the btree. I think it gets to the right leaf now.

But then at this point:
https://github.com/memecode/apfs-tools/commit/36d0cee109df6c7c344d63955f8eae7c21e3ae21#diff-ed645dbf88753c45bcd7100d5e899dd3R420

It's finding the array is fully traversed and exits the loop before collecting the records. Sigh.

**Comment by memecode** [comment]
Ok ok ok... more progress:
https://github.com/memecode/apfs-tools/commit/29ba561096fc87a931a516aa1fb9e374554a375d

Now I can get the fs records for the directory that was failing!!! Wow... these changes actually work for me. I was so close with the last post. Seems to be an issue with very recent OID's... the very highest numbers. Which would of course be on the very edge of the tree.

**Comment by jivanpal** [comment]
The `master` branch does not contain some fixes to the B-tree implementation that I made some time ago in the `unstable`, but have not yet had the time to merge. See here: https://github.com/jivanpal/apfs-tools/blob/unstable/src/apfs/func/btree.h#L343

**Comment by jivanpal** [comment]
I greatly appreciate the effort you are putting in, but at this point, it is probably doubled (and thus, in a sense, wasted) effort. I need to go through the codebase and reconcile quite a lot of working changes that I made when I was just using this toolset to solve my own personal data loss issues and those of a few people who directly approached me. I will be spending the next few weeks consolidating everything, versioning the toolset, and documenting known issues on GitHub. After that, your contributions will probably be more useful.

In the meantime, if you want to see the latest version of the code, please work on the `unstable` branch. Thank you for your understanding.

**Comment by memecode** [comment]
Well as you can probably guess I also have my own data loss issues that need to be resolved "yesterday" hence the amount of effort I'm putting in. I'll have a look at the unstable branch and see what I can do to merge those btree changes into what I'm doing. But I don't have weeks to wait for your refactoring to get these files off my corrupt disk.

And likely once I have those files, and I start back full time work post covid redundancy, my interest and availability to work on the code base will drop off. At least those the parsing of the INODE x fields to get the exact file size will be a great addition.

**Comment by jivanpal** [comment]
Fair enough 🙂 If you're able to recover your files using what's in `unstable`, then `/supplemental-tools/remove-trailing-nulls.sh` should get you sorted for now.

**Comment by memecode** [comment]
> Then `/supplemental-tools/remove-trailing-nulls.sh` should get you sorted for now.

I had that "remove-trailing-nulls" functionality built into my python script and it doesn't work for the Logic Pro X projects that I have. Hence adding the code to pull the correct file size. Because I actually really needed that.

**Comment by memecode** [comment]
I think for the most part your unstable changes work. And the combination of our fixes in my fork definitely address this. I got 100% coverage on my home folder (100k+ files) in terms of listing directories. Some of the files couldn't be exported by the apfs-recover though. In my case they weren't important ones so I'm not digging into exactly what went wrong there.

**Comment by jivanpal** [comment]
Finally got round to resuming work on this... I've rewritten the Git history to a new branch,`main`, which breaks down what was previously amassed in a couple of commits in `unstable` — it was very difficult to see what I did upon coming back to this, haha!

> I think for the most part your unstable changes work.

Are you able to recall or find out what changes you made that improve upon what's in `unstable`? I think my B-tree implementation works without issue, and from what I can see, we've effectively made the same changes. Would be good to know if there was anything else you changed; but no worries if you don't have time for that, as I'll be checking for bugs anyway.

Closing this for now.
